pixel mc gain calibration scheme: new GTs for MC Run3, modified Clusterizer conditions for Run3

dkotlins opened the pull request:
#### PR description:

<!--   -->
Modify the scheme of the pixel gain calibration. Instead of applying the vcal calibration in the 
clusterizer include it already in the gain calibration payload.  Like this the vcal calibration can 
be easily adjusted for each new gain calibration.
This is the first step in which we only change the MC gain calibrations. It should apply to any future Run3 simulations. Data from Run1/2 and older, none Run3 MC, will not be affected.

To be applied for Run 3 MC & Data. No code changes, the only changes are:
1. new GTs for Run3 MC which include the new gain calibration values,
2. changes in the clusterizer parameters for Run 3, for the VCAL calibration use slope=1 and offset =0.


  - what changes are expected in the output if any,
No changes expected to the output 
  - what other PRs or externals it depends upon if any,
There are no other PRs, but the change must be in sync with a GT change, this is why the autoCond.py file is also modified.

  - link to any additional material useful to provide a documentation for this PR (slides, JIRA tickets, related pull requestes, hypernews, TWiki or Indico pages)  
https://indico.cern.ch/event/879470/contributions/3796405/attachments/2009273/3356603/pix_off_25_3_gain_calibration_mc.pdf

#### PR validation:

<!-- Normal, standard testing, no additions no updates to the runTheMatrix test workflows -->

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No backport required.

<!-- Please replace this text with any link to  -->

The changes have been verified using Run3 simulations.

dkotlins committed:
change the pixel mc gain calibration scheme

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14411

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py  modified in PR(s): #27983


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14411/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14411/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

dkotlins committed:
fix a typo in 2024 tag

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14420

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/AlCa/python/autoCond.py  modified in PR(s): #29230
  - File  RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py  modified in PR(s): #27983


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14420/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14420/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

mmusich commented:
@dkotlins this PR has code style and quality issues. Can you please resolve them as suggested at https://github.com/cms-sw/cmssw/pull/29333#issuecomment-605903814? 
In alternative you can run `scram b code-format` and `scram b code-checks` in your local development area and then push the changes.
Otherwise the PR integration tests cannot be performed.

dkotlins committed:
fix formatting issues, use scram b code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14422

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  Configuration/AlCa/python/autoCond.py  modified in PR(s): #29230
  - File  RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py  modified in PR(s): #27983

cmsbuild commented:
A new Pull Request was created by  @dkotlins (danek) for master.

It involves the following packages:

CondTools/SiPixel
Configuration/AlCa
RecoLocalTracker/SiPixelClusterizer


@perrotta, @slava77, @christopheralanwest, @tocheng, @cmsbuild, @tlampen, @ggovi, @pohsun can you please review it and eventually sign? Thanks.
@fabiocos, @makortel, @felicepantaleo, @GiacomoSguazzoni, @rovere, @VinInn, @Martin-Grunewald, @tocheng, @gpetruc, @mmusich, @threus, @ebrondol this is something you requested to watch as well.
@davidlange6, @silviodonato, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9a30a155de7aa641f81f721999553d0b1b519660
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5bdafd/5430/summary.html
CMSSW: CMSSW_11_1_X_2020-03-29-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

mmusich reviewed (commented):
@dkotlins I have few comments, mostly stylistic.

mmusich commented on Configuration/AlCa/python/autoCond.py:
```diff
@@ -65,19 +65,19 @@
+    'phase1_2023_realistic'    : '111X_mcRun3_2023_realistic_newPixelGain_v1', # GT containing realistic conditions for Phase1 2023
     # GlobalTag for MC production with realistic conditions for Phase1 2024
-    'phase1_2024_realistic'    : '111X_mcRun3_2024_realistic_v3', # GT containing realistic conditions for Phase1 2024
+    'phase1_2024_realistic'    : '111X_mcRun3_2024_realistic_newPixelGain_v1', # GT containing realistic conditions for Phase1 2024
```
@christopheralanwest @tlampen @tocheng given that these Global Tags are meant to be integrated in release rather than for testing, shouldn't they named according to the normal naming conventions?

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5bdafd/5430/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3775 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 44572
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2647219
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

dkotlins commented:
Concerning the Era=Run3 modification, I thought that it was necessary to define it again.
This constants might be anyway modified for Run3. 
I will remove it now.
D.

> On 30 Mar 2020, at 15:40, Marco Musich <notifications@github.com> wrote:
> 
> 
> @mmusich commented on this pull request.
> 
> @dkotlins I have few comments, mostly stylistic.
> 
> In CondTools/SiPixel/test/SiPixelCondObjBuilder.cc:
> 
> > @@ -40,7 +51,7 @@ namespace cms {
>      uint32_t nchannels = 0;
>      //    int mycol = 415;
>      //    int myrow = 159;
> -
> +    std::cout<<electronsPerVcal_<<" "<<electronsPerVcal_Offset_<<" "<<electronsPerVcal_L1_<<" "<<electronsPerVcal_L1_Offset_<<std::endl;
> 
> cout is not thread safe, therefore discouraged. As this class is a test directory it can be tolerated, though it would preferable to make it uniform with the other logging practices (especially as I see that LogInfo is already extensively used her).
> 
> In RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py:
> 
> > @@ -17,6 +16,22 @@
>    ClusterThreshold_L1     = 2000
>  )
>  
> +# Run3, changes in the gain calibration scheme 
> +#from Configuration.Eras.Era_Run3_cff import Run3
> +#Run3.toModify(siPixelClusters,
> +from Configuration.Eras.Modifier_run3_common_cff import run3_common
> +run3_common.toModify(siPixelClusters,
> +  VCaltoElectronGain      = 1,  # all gains=1, pedestals=0
> +  VCaltoElectronGain_L1   = 1,  # 
> +  VCaltoElectronOffset    = 0,  # 
> +  VCaltoElectronOffset_L1 = 0,  # 
> +  ChannelThreshold        = 10,
> 
> why the following four parameters: ChannelThreshold, SeedThreshold , ClusterThreshold and ClusterThreshold_L1 are specified again in this modifier?
> That is unnecessary as the Run3 era includes the phase1Pixel modifier already, hence those should be already correctly set.
> 
> In CondTools/SiPixel/test/SiPixelCondObjBuilder.cc:
> 
> > +	    //gain += secondRocRowGainOffset_;
> +
> +           // include the vcal claibration already here 
> +
> +           double newGain = 1, newPed  = 0.;
> +	   if(layer==1) { 
> +	     newGain = gain * electronsPerVcal_L1_;
> +	     newPed  = ped  - (electronsPerVcal_L1_Offset_/newGain);
> +	   } else {
> +	     newGain = gain * electronsPerVcal_;
> +	     newPed  = ped  - (electronsPerVcal_Offset_/newGain);
> +	   }
> +	   ped = newPed;
> +	   gain = newGain;
> +	   
> +	   if (gain > maxgain) gain = maxgain;
> 
> std::clamp here https://en.cppreference.com/w/cpp/algorithm/clamp ?
> 
> In CondTools/SiPixel/test/SiPixelCondObjBuilder.cc:
> 
> > +	   if(layer==1) { 
> +	     newGain = gain * electronsPerVcal_L1_;
> +	     newPed  = ped  - (electronsPerVcal_L1_Offset_/newGain);
> +	   } else {
> +	     newGain = gain * electronsPerVcal_;
> +	     newPed  = ped  - (electronsPerVcal_Offset_/newGain);
> +	   }
> +	   ped = newPed;
> +	   gain = newGain;
> +	   
> +	   if (gain > maxgain) gain = maxgain;
> +	   else if (gain < mingain) gain = mingain;
> +	   if (ped > maxped) ped = maxped;
> +	   else if (ped < minped) ped = minped;
> +      
> +	   //std::cout<<gain<<" "<<ped<<" "<<newGain<<" "<<newPed<<std::endl;
> 
> why this commented code?
> 
> In CondTools/SiPixel/test/SiPixelCondObjForHLTBuilder.cc:
> 
> >              }
> +            ped = newPed;
> +            gain = newGain;
> +
> +	    if (gain > maxgain) gain = maxgain;
> 
> std::clamp here?
> 
> In CondTools/SiPixel/test/SiPixelCondObjOfflineBuilder.cc:
> 
> > -                ped = minped;
> -            }
> +	 
> +	    // include the vcal claibration already here 
> +	    double newGain = 1, newPed  = 0.;
> +	    if(layer==1) { 
> +	      newGain = gain * electronsPerVcal_L1_;
> +	      newPed  = ped  - (electronsPerVcal_L1_Offset_/newGain);
> +	    } else {
> +	      newGain = gain * electronsPerVcal_;
> +	      newPed  = ped  - (electronsPerVcal_Offset_/newGain);
> +	    }
> +	    ped = newPed;
> +	    gain = newGain;
> +
> +	    if (gain > maxgain) gain = maxgain;
> 
> std::clamp here?
> 
> In CondTools/SiPixel/test/SiPixelCondObjBuilder.cc:
> 
> >      SiPixelGainCalibration_ = new SiPixelGainCalibration(minped, maxped, mingain, maxgain);
>  
>      edm::ESHandle<TrackerGeometry> pDD;
>      iSetup.get<TrackerDigiGeometryRecord>().get(pDD);
>      edm::LogInfo("SiPixelCondObjBuilder") << " There are " << pDD->dets().size() << " detectors" << std::endl;
> +   //Retrieve tracker topology from geometry
> +    edm::ESHandle<TrackerTopology> tTopo;
> +    iSetup.get<TrackerTopologyRcd>().get(tTopo);
> +    //const TrackerTopology* tt = tTopo.product();
> 
> unneeded commented code.
> 
> In CondTools/SiPixel/test/SiPixelCondObjBuilder.cc:
> 
> > @@ -67,6 +84,12 @@ namespace cms {
>  
>          const PixelGeomDetUnit* pixDet = dynamic_cast<const PixelGeomDetUnit*>((*it));
>          const PixelTopology& topol = pixDet->specificTopology();
> +
> +	// Find out if it is layer 1
> +        //DetId detId = (*it)->geographicalId();
> 
> unneeded commented code.
> 
> In CondTools/SiPixel/test/SiPixelCondObjForHLTBuilder.cc:
> 
> >      SiPixelGainCalibration_ = new SiPixelGainCalibrationForHLT(minped, maxped, mingain, maxgain);
>  
>      edm::ESHandle<TrackerGeometry> pDD;
>      iSetup.get<TrackerDigiGeometryRecord>().get(pDD);
>      edm::LogInfo("SiPixelCondObjForHLTBuilder") << " There are " << pDD->dets().size() << " detectors" << std::endl;
> +   //Retrieve tracker topology from geometry
> +    edm::ESHandle<TrackerTopology> tTopo;
> +    iSetup.get<TrackerTopologyRcd>().get(tTopo);
> +    //const TrackerTopology* tt = tTopo.product();
> 
> unneeded commented code.
> 
> In CondTools/SiPixel/test/SiPixelCondObjForHLTBuilder.cc:
> 
> > @@ -81,6 +92,11 @@ namespace cms {
>  
>          const PixelGeomDetUnit* pixDet = dynamic_cast<const PixelGeomDetUnit*>((*it));
>          const PixelTopology& topol = pixDet->specificTopology();
> +      // Find out if it is layer 1
> +        //DetId detId = (*it)->geographicalId();
> 
> unneeded commented code.
> 
> In CondTools/SiPixel/test/SiPixelCondObjOfflineBuilder.cc:
> 
> >      SiPixelGainCalibration_ = new SiPixelGainCalibrationOffline(minped, maxped, mingain, maxgain);
>  
>      edm::ESHandle<TrackerGeometry> pDD;
>      iSetup.get<TrackerDigiGeometryRecord>().get(pDD);
>      edm::LogInfo("SiPixelCondObjOfflineBuilder") << " There are " << pDD->dets().size() << " detectors" << std::endl;
>  
> +
> +    //Retrieve tracker topology from geometry
> +    edm::ESHandle<TrackerTopology> tTopo;
> +    iSetup.get<TrackerTopologyRcd>().get(tTopo);
> +    //const TrackerTopology* tt = tTopo.product();
> 
> unneeded commented code.
> 
> In CondTools/SiPixel/test/SiPixelCondObjOfflineBuilder.cc:
> 
> > @@ -79,6 +95,12 @@ namespace cms {
>  
>          const PixelGeomDetUnit* pixDet = dynamic_cast<const PixelGeomDetUnit*>((*it));
>          const PixelTopology& topol = pixDet->specificTopology();
> +
> +	// Find out if it is layer 1
> +	//DetId detId = (*it)->geographicalId();
> 
> unneeded commented code.
> 
> In Configuration/AlCa/python/autoCond.py:
> 
> > @@ -65,19 +65,19 @@
>      # GlobalTag for MC production (cosmics) with realistic conditions for full Phase1 2018 detector,  Strip tracker in DECO mode
>      'phase1_2018_cosmics'      :   '110X_upgrade2018cosmics_realistic_deco_v6',
>      # GlobalTag for MC production (cosmics) with realistic conditions for full Phase1 2018 detector,  Strip tracker in PEAK mode
> -    'phase1_2018_cosmics_peak' :   '110X_upgrade2018cosmics_realistic_peak_v7',
> +    'phase1_2018_cosmics_peak' :   '110X_upgrade2018cosmics_realistic_peak_v7',   
> 
> remove extra space.
> 
> In CondTools/SiPixel/test/SiPixelCondObjBuilder.cc:
> 
> >  
> -            // 	   gain =  2.8;
> -            // 	   ped  = 28.2;
> +            double newGain = 1, newPed = 0.;
> +            if (layer == 1) {
> +              newGain = gain * electronsPerVcal_L1_;
> +              newPed = ped - (electronsPerVcal_L1_Offset_ / newGain);
> +            } else {
> +              newGain = gain * electronsPerVcal_;
> +              newPed = ped - (electronsPerVcal_Offset_ / newGain);
> +            }
> +            ped = newPed;
> +            gain = newGain;
> +
> +            if (gain > maxgain)
> 
> std::clamp here? https://en.cppreference.com/w/cpp/algorithm/clamp
> 
> In CondTools/SiPixel/test/SiPixelCondObjForHLTBuilder.cc:
> 
> >              }
> +            ped = newPed;
> +            gain = newGain;
> +
> +            if (gain > maxgain)
> 
> std::clamp here?
> 
> In CondTools/SiPixel/test/SiPixelCondObjOfflineBuilder.cc:
> 
> > @@ -79,6 +90,12 @@ namespace cms {
>  
>          const PixelGeomDetUnit* pixDet = dynamic_cast<const PixelGeomDetUnit*>((*it));
>          const PixelTopology& topol = pixDet->specificTopology();
> +
> +        // Find out if it is layer 1
> +        //DetId detId = (*it)->geographicalId();
> 
> remove unneeded commented code.
> 
> In CondTools/SiPixel/test/SiPixelCondObjOfflineBuilder.cc:
> 
> >              }
> +            ped = newPed;
> +            gain = newGain;
> +
> +            if (gain > maxgain)
> 
> std::clamp here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

dkotlins commented:
Here I am not guilty. These names were provided by Tongguang Cheng.
I was just trying to make a consistent PR, one which works, but after all it is not my 
task to find good names for GTs.

I propose to cancel this Pull request. 
I will make a new one which will include just the pixel change and I will specify 
that it has to be accompanied by a corresponding new GT.
The two will have to be synchronised, otherwise it will not work.
D.

> On 30 Mar 2020, at 15:41, Marco Musich <notifications@github.com> wrote:
> 
> 
> @mmusich commented on this pull request.
> 
> In Configuration/AlCa/python/autoCond.py:
> 
> >      # GlobalTag for MC production with realistic conditions for Phase1 2024
> -    'phase1_2024_realistic'    : '111X_mcRun3_2024_realistic_v3', # GT containing realistic conditions for Phase1 2024
> +    'phase1_2024_realistic'    : '111X_mcRun3_2024_realistic_newPixelGain_v1', # GT containing realistic conditions for Phase1 2024
> 
> @christopheralanwest @tlampen @tocheng given that these Global Tags are meant to be integrated in release rather than for testing, shouldn't they named according to the normal naming conventions?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

mmusich commented:
> I propose to cancel this Pull request.
I will make a new one which will include just the pixel change and I will specify
that it has to be accompanied by a corresponding new GT.
The two will have to be synchronised, otherwise it will not work.

I'd just rather wait for AlCa/DB to clarify if they plan to provide better named GTs or they plan to sign the update as is.

tocheng commented:
I will give GTs official names and you can update them with a commit. Thanks.

tocheng commented:
@dkotlins @mmusich 
Please use this commit
https://github.com/tocheng/cmssw/commit/1f941d72f1b7d06d226f80328b0c2dfbb901b0ae

Thanks.

dkotlins commented:
I am not sure what should I do.
Can I use your commit somehow or should I just retype your names to my commit?
D.

> On 1 Apr 2020, at 17:52, Tongguang <notifications@github.com> wrote:
> 
> 
> @dkotlins @mmusich
> Please use this commit
> tocheng@1f941d7
> 
> Thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

mmusich commented:
@dkotlins Danek, for you it is probably faster just to retype the new names in your local area and then push again. Please try to address also the other comments.
Thanks
M.

dkotlins committed:
switch to the new GTs, take out the CondTools code

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29333/14485

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py  modified in PR(s): #27983

cmsbuild commented:
Pull request #29333 was updated. @perrotta, @slava77, @christopheralanwest, @tocheng, @cmsbuild, @tlampen, @pohsun can you please check and sign again.

christopheralanwest commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3935d7ee4d27e3954c306ff76dad62bbf499d468
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5bdafd/5504/summary.html
CMSSW: CMSSW_11_1_X_2020-04-02-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-5bdafd/5504/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3486 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2692110
 - DQMHistoTests: Total failures: 38879
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2652912
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

perrotta commented:
@dkotlins (@mmusich) please edit the title and the description of this PR so that it can reflect what is done here now that you removed all the CondTools related part:
- new GTs
- skip the vcal calibration in the clusterizer, by setting "slope = 1 & offfset = 0." in the vcal

Please also check the differences in the Run3 outputs, and verify whether they correspond to what expected

dkotlins commented:
**PR: Description**

New scheme for the pixel gain calibration.
To be applied for Run 3 MC & Data. No code changes.
- new GTs for Run3 MC which include the new gain calibration values,
- changes in the clusterizer parameters for Run 3, for the VCAL calibration use slope=1 and offset =0.

dkotlins commented:
I have changed the title.
I do not know how to modify the initial PR description, so I just added a comment.

I could not find any relevant changes coming from this PR. 
There are all sorts of ECAL/HCAL differences but I do not see how this is 
related to pixels.
The changes in the pixel cluster charge are tiny, below any statistical significance.

D.
> On 3 Apr 2020, at 12:13, perrotta <notifications@github.com> wrote:
> 
> 
> @dkotlins (@mmusich) please edit the title and the description of this PR so that it can reflect what is done here now that you removed all the CondTools related part:
> 
> 	• new GTs
> 	• skip the vcal calibration in the clusterizer, by setting "slope = 1 & offfset = 0." in the vcal
> Please also check the differences in the Run3 outputs, and verify whether they correspond to what expected
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

perrotta commented:
![pr](https://user-images.githubusercontent.com/4069749/78543187-95f9eb00-77f8-11ea-8196-44f2c96d5552.jpg)

In the top right of the PR description panel you can access the commands for it. One of them is "edit", which is only allowed to the owner of the PR

perrotta commented:
There are several pixel related differences in jenkins outputs, which I assume due to the modified GT.
I imagine that those differences propagate with some tiny effect to the reconstructed quantities, but I think it is worth to validate/confirm in some larger sample that the low-stat jenkins tests outputs.

By the way, for my own knowledge, could you please point out were the calibrations taken from the DB replaces in run3 workflows the ones set by the gain/offset parameters, which are set here to 0/1 respectively?

dkotlins commented:
I am not sure that I understand the question.
Until now the gain calibration was taken from the DB and than corrected for the vcal calibration
in the clusterizer using parameters taken from the configuration.
Now the vcal calibration constants are already inlcuded in the gain calibration in the DB.
So the constant taken form the DB do not need it. All the calibrations are included in the DN payload.

This is explained in slides, link below, there are also plots there with higher statistics showing that the 
two approaches give the same answers:
https://indico.cern.ch/event/879470/contributions/3796405/attachments/2009273/3356603/pix_off_25_3_gain_calibration_mc.pdf

The results cannot be exactly the same because of random generators using for smearing around the average values 
and probably because of binning effects . 
The 2nd one is because if:
 - you take a distribution and bin it, 
  - take the same distribution, rescale it, bin it, rescale back,
I believe the 2 will have some numerical differences.

D.

> On 6 Apr 2020, at 11:38, perrotta <notifications@github.com> wrote:
> 
> 
> There are several pixel related differences in jenkins outputs, which I assume due to the modified GT.
> I imagine that those differences propagate with some tiny effect to the reconstructed quantities, but I think it is worth to validate/confirm in some larger sample that the low-stat jenkins tests outputs.
> 
> By the way, for my own knowledge, could you please point out were the calibrations taken from the DB replaces in run3 workflows the ones set by the gain/offset parameters, which are set here to 0/1 respectively?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

perrotta commented:
@dkotlins sorry for getting late on it, as I was busy with other things.

My question starts from 
"For this to work one has to skip the vcal calibration in the clusterizer.
No code changes are need, it is enough to set the vcal to
slope = 1 & offfset = 0."

Questions:
- Do I understand correctly that with such a "neutralization" of slope and offset in the python configurations those values are taken from the DB, while otherwise the values written in the python configuration are taken instead?
- Could you please point me the place in the code where such a switch to the GT given those values for slope and offset is actually implemented?
- Why the same procedure of having gains and offsets in the DB is not implemented also for Run1/Run2? Is it just because you consider them fozen, and you don't consider there can be further calibrations that can modify them?

mmusich commented:
@perrotta taking the liberty to provide some answers, @dkotlins can of course complement.

 >  Do I understand correctly that with such a "neutralization" of slope and offset in the python configurations those values are taken from the DB, while otherwise the values written in the python configuration are taken instead?

The per pixel gains are still taken from DB, only now the values don't get multipied and offset because that operation is already done *within* the numbers stored in DB.

>    Could you please point me the place in the code where such a switch to the GT given those values for slope and offset is actually implemented?

There is no such switch, the gain payloads for Run3 in the DB are just populated differently, taking into account that slope is 1 and offset is 0, so the "new" gains get multiplied by a factor ~ 50 see plot below (red is "old" gain scheme, blue is "new gain scheme).
The net effect is that the calibration remains ~ the same (with the caveat that Danek reported at https://github.com/cms-sw/cmssw/pull/29333#issuecomment-610206141)

![image](https://user-images.githubusercontent.com/5082376/78984444-8515e600-7b26-11ea-9b02-6b14a8aa335b.png)
 
>    Why the same procedure of having gains and offsets in the DB is not implemented also for Run1/Run2? Is it just because you consider them fozen, and you don't consider there can be further calibrations that can modify them?

To avoid redoing dozens of payloads to keep physics the same... 
The run2 / run3 boundary is an excellent moment to do this, because in this way we can just switch to the new scheme for the future payloads (all of them, of course, yet to be derived).

perrotta commented:
+1
- For reco, only the SiPixelClusterizer_cfi.py configuration gets modified for Run3, where "neutral" values are assigned to slope and offset so that the pixel mc gain remains only defined by the payloads in the GT
- These new gains for Run3 are somehow modified with respect to those derived from the previous formulation. The differences observed in the jenkins outputs are minor, but probably a bit larger than possible numerical rounding effects, and they propagate to higher level quantities as reduced calo recHits, electrons, jets, met, etc.
  - AlCa will take care of verifying those new gains in the modified GT payloads while signing
  - A clearer picture can be obtained with the usual pre-release validations: I look forward this PR can enter the next 11_X pre-release, so that its effects can be thoroughly inspected

mmusich commented:
Can AlCa please review?
@tlampen @tocheng @christopheralanwest

tocheng commented:
+1

Sorry for being late.

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
the differences in DQM plots are explained in https://github.com/cms-sw/cmssw/pull/29333#issuecomment-610206141
